### PR TITLE
Feat/signer id page

### PR DESCRIPTION
**Pull request by BLuEScioN**
## What type of PR is this? (check all applicable)
- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Adds the signer id page 
We are waiting on the API for the data for the stats panel on the right 

Figma - https://www.figma.com/design/31q1g27k69z1yXb6r6vrRD/Stacks-Explorer-Working-File?node-id=9664-20852&node-type=canvas&t=PVJNUe8s2skW3xi9-0

## Issue ticket number and link
https://github.com/hirosystems/explorer/issues/1878

# Checklist:
- [X] I have performed a self-review of my code.
- [X] I have tested the change on desktop and mobile.
- [X] I have added thorough tests where applicable.
- [X] I've added analytics and error logging where applicable.

### CODE REVIEW

File: src/app/signers/SignersTable.tsx [hunk 1]
```diff
@@ -305,48 +283,63 @@ const SignersTableBase = () => {
...
-  );
+  const signersMetrics =
+    useInfiniteQueryResult<SignerMetricsSignerForCycle>(signersMetricsResponse);
+
+  const signersData = useMemo(() => {
+    return signers
+      .map(signer => {
+        const metrics =
+          signersMetrics.find(m => m.signer_key === signer.signing_key) ||
+          ({} as SignerMetricsSignerForCycle);
```
Comment (He1DAr) [inline src/app/signers/SignersTable.tsx:307]: If metrics is passed as empty object to formatSignerRowData it'll cause totalProposals to be NaN

File: src/app/signers/data/UseSigner.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  stacked_amount_percent: number;
+  solo_stacker_count: number;
+  pooled_stacker_count: number;
+}
+
+export interface StackingHistoryInfo extends SignerMetricsSignerForCycle {
+  cycleid: number;
+}
+
+interface SignerError {
```
Comment (He1DAr) [inline src/app/signers/data/UseSigner.ts:34]: not used anywhere

File: src/app/signers/data/UseSigner.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      queries,
+      combine: (response: UseQueryResult<SignerMetricsSignerForCycle, Error>[]) => {
+        return response.map(r => r.data);
+      },
+    };
+  }, [queries]);
+  const signerStackingHistory = useQueries(signerStackingHistoryQueries, queryClient);
+
+  const found404s = useMemo(() => {
+    return signerStackingHistory.some(r => {
```
Comment (He1DAr) [inline src/app/signers/data/UseSigner.ts:82]: I suggest using a descriptive name for this variable

File: src/app/signers/data/UseSigner.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  const getQuery = useGetSignerMetricsSignerForCycleQuery();
+  const [offset, setOffset] = useState(0);
+  const cyclesToQuery = useMemo(() => {
+    return selectedCycle
+      ? [parseInt(selectedCycle)]
+      : Array.from({ length: offset + DEFAULT_LIST_LIMIT }, (_, i) => currentCycleId - i);
+  }, [selectedCycle, currentCycleId, offset]);
+  const queryClient = useQueryClient();
+  const queries = useMemo(() => {
+    return cyclesToQuery.map(cycle => {
```
Comment (He1DAr) [inline src/app/signers/data/UseSigner.ts:66]: It looks like this part will do multiple queries for all cycles? If that's the case I think it could cause problems with spawning many requests and spamming the API

File: src/app/signers/data/UseSigner.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      combine: (response: UseQueryResult<SignerMetricsSignerForCycle, Error>[]) => {
+        return response.map(r => r.data);
+      },
+    };
+  }, [queries]);
+  const signerStackingHistory = useQueries(signerStackingHistoryQueries, queryClient);
+
+  const found404s = useMemo(() => {
+    return signerStackingHistory.some(r => {
+      if (r && 'error') {
```
Comment (He1DAr) [inline src/app/signers/data/UseSigner.ts:83]: the expression 'error' is always truthy

File: src/app/signers/data/UseSigner.ts (hunk 3 above)
Comment (He1DAr) [inline src/app/signers/data/UseSigner.ts:82]: could you explain why if signerStackingHistory has any entities that means there's 404 error?

File: src/app/signers/data/UseSignerAddresses.ts [hunk 6]
```diff
@@ -1,21 +1,31 @@
 // Add missing import statement
```
Comment (He1DAr) [inline src/app/signers/data/UseSignerAddresses.ts:1]:
```suggestion
```

File: src/app/signers/data/UseSignerAddresses.ts [hunk 7]
```diff
@@ -27,15 +37,59 @@ export function useGetStackersBySignerQuery() {
...
+  return response.json();
+};
+
+export function useSuspenseSignerStackersInfinite(
+  cycleId: number,
+  signerKey: string,
+  options: any = {}
+): UseInfiniteQueryResult<InfiniteData<GenericResponseType<SignersStackersData>>> {
+  // const api = useApi();
+  // if (!address) throw new Error('Address is required');
```
Comment (He1DAr) [inline src/app/signers/data/UseSignerAddresses.ts:82]:
```suggestion
```

File: src/app/signers/data/UseSignerAddresses.ts [hunk 8]
```diff
@@ -27,15 +37,59 @@ export function useGetStackersBySignerQuery() {
...
+  const { url: activeNetworkUrl } = useGlobalContext().activeNetwork;
+
+  return useInfiniteQuery<GenericResponseType<SignersStackersData>>({
+    queryKey: [SIGNER_STACKERS_INFINITE_QUERY_KEY, cycleId, signerKey],
```
Comment (He1DAr) [inline src/app/signers/data/UseSignerAddresses.ts:86]: should options be added to query key so e.g. different limit would trigger a different request?

File: src/app/signers/data/signer-metrics-hooks.ts [hunk 9]
```diff
@@ -90,15 +91,30 @@ const fetchSignersForCycle = async (
   return response.json();
 };
 
+export function useSuspensePoxSigners(cycleId: string) {
```
Comment (He1DAr) [inline src/app/signers/data/signer-metrics-hooks.ts:94]: not used anywhere

File: src/app/signers/data/useSigners.ts [hunk 10]
```diff
@@ -13,19 +13,42 @@ export interface SignerInfo {
...
-    Error,
-    ApiResponseWithResultsOffset<SignerInfo>
-  >({
-    queryKey: ['signers', cycleId],
-    queryFn: () =>
-      fetch(`${activeNetworkUrl}/extended/v2/pox/cycles/${cycleId}/signers`).then(
-        res => res.json() as Promise<ApiResponseWithResultsOffset<SignerInfo>>
-      ),
+  return useSuspenseInfiniteQuery<ApiResponseWithResultsOffset<PoxSigner>>({
+    queryKey: [POX_SIGNERS_QUERY_KEY, cycleId],
```
Comment (He1DAr) [inline src/app/signers/data/useSigners.ts:46]: same question around adding options to query key

**Comment by He1DAr** [comment]
image
I think copy button here should copy stacker_address rather than pox_address

**Comment by He1DAr** [comment]
image
we might wanna shrink this extra white space

**Comment by He1DAr** [comment]
there's a bit of screen jump when clicking cycle arrow button

https://github.com/user-attachments/assets/bbffbb78-2d91-454a-9ab0-4010792d026e

**Comment by He1DAr** [comment]
image
we might wanna make these cards same width on smaller screen (unless this is intentional)

**Comment by He1DAr** [comment]
image
maybe we truncate the signer key even more on mobile to allow more width to the table?

File: src/app/signer/[signerKey]/AssociatedAddressesTable.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+import { Stack } from '../../../ui/Stack';
+import { Text } from '../../../ui/Text';
+import { ExplorerErrorBoundary } from '../../_components/ErrorBoundary';
+import { useSuspenseCurrentStackingCycle } from '../../_components/Stats/CurrentStackingCycle/useCurrentStackingCycle';
+import { AssociatedAddressesTableSkeleton } from './skeleton';
+
+export const addressItemHeight = 69;
+
+export const AssociatedAddressesTableLayout = ({ children }: { children: ReactNode }) => {
+  return <Section title="Associated addresses">{children}</Section>;
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/AssociatedAddressesTable.tsx:21]: The wrapper is in case this ever changes. The wrapper is used for the skeleton to keep it in sync with the component

File: src/app/signer/[signerKey]/CycleSortFilter.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import { useMemo } from 'react';
+
+import { FilterMenu } from '../../../common/components/FilterMenu';
+
+export enum CycleSortOrder {
+  Asc = 'asc',
+  Desc = 'desc',
+}
+
+function getSortOptionLabel(order: CycleSortOrder) {
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/CycleSortFilter.tsx:11]: ok

File: src/app/signer/[signerKey]/PageClient.tsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import { Grid, GridProps } from '../../../ui/Grid';
+import { Stack } from '../../../ui/Stack';
+import { PageTitle } from '../../_components/PageTitle';
+import { AssociatedAddressesTable } from './AssociatedAddressesTable';
+import { SignerStats } from './SignerStats';
+import { SignerSummary } from './SignerSummary';
+import { StackingHistoryTable } from './StackingHistoryTable';
+
+export function SignerKeyPageLayout(props: GridProps) {
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/PageClient.tsx:13]: These layouts are here to make sure that the page/component skeletons stay in sync.

File: src/app/signer/[signerKey]/SignerStats.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+import { Stack } from '../../../ui/Stack';
+import { Caption } from '../../../ui/typography';
+import { useSuspenseCurrentStackingCycle } from '../../_components/Stats/CurrentStackingCycle/useCurrentStackingCycle';
+import { SignerKeyStatsSkeleton } from './skeleton';
+
+interface SignerStatsProps {
+  signerKey: string;
+}
+
+export const SignerStatsLayout = ({ children }: { children: ReactNode }) => {
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/SignerStats.tsx:17]: Again. This layout is important for the component skeleton.

File: src/app/signer/[signerKey]/SignerSummary.tsx [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+import { useSuspenseCurrentStackingCycle } from '../../_components/Stats/CurrentStackingCycle/useCurrentStackingCycle';
+import { getEntityName } from '../../signers/SignersTable';
+import { useSuspensePoxSigner } from '../../signers/data/UseSigner';
+import { SignerSummarySkeleton } from './skeleton';
+
+interface SignerSummaryProps {
+  signerKey: string;
+}
+
+export const SignerSummaryLayout = ({ children }: { children: ReactNode }) => {
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/SignerSummary.tsx:18]: It's not single use. This layout is important for the component skeleton.

File: src/app/signer/[signerKey]/SignerSummary.tsx [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    <Section title="Summary">
+      <Stack width="100%">{children}</Stack>
+    </Section>
+  );
+};
+
+export const SignerSummaryBase = ({ signerKey }: SignerSummaryProps) => {
+  const { currentCycleId } = useSuspenseCurrentStackingCycle();
+  const { data: signerData } = useSuspensePoxSigner(currentCycleId, signerKey);
+  const numAssociatedAddresses = signerData?.solo_stacker_count + signerData?.pooled_stacker_count;
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/SignerSummary.tsx:29]: ok

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+    latency: singerInfo.average_response_time_ms ? singerInfo.average_response_time_ms : 50,
+    approved: singerInfo.proposals_accepted_count
+      ? singerInfo.proposals_accepted_count / totalProposals
+      : 0.85,
+    rejected: singerInfo.proposals_rejected_count
+      ? singerInfo.proposals_rejected_count / totalProposals
+      : 0.1,
+    missing: singerInfo.proposals_missed_count
+      ? singerInfo.proposals_missed_count / totalProposals
+      : 0.05,
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:218]: this was test code. there are no defaults

File: src/app/signers/data/UseSigner.ts (hunk 4 above)
Comment (BLuEScioN) [reply src/app/signers/data/UseSigner.ts:66]: There isn't a batch endpoint, so I don't have any choice. Here is the endpoint that I am using to get signer data for a signer for a given cycle - https://signer-metrics-ck5uyle9e-hirosystems.vercel.app/#tag/Signers/operation/getPoxCycleSigner. As you can see I limit the queries to 5, which I think is an acceptable # of requests to make. Sure the user can spam this until there are no more cycles, but they could also do that on any of our other pages, and besides there aren't even 5 cycles to query for currently. 

TLDR: To resolve this we need a batch endpoint, but I don't think this is too bad, considering we've done similar and worse elsewhere

File: src/app/signers/data/UseSigner.ts (hunk 3 above)
Comment (BLuEScioN) [reply src/app/signers/data/UseSigner.ts:82]: ok

Comment (BLuEScioN) [reply src/app/signers/data/UseSigner.ts:82]: done.

File: src/app/signers/data/UseSigner.ts (hunk 5 above)
Comment (BLuEScioN) [reply src/app/signers/data/UseSigner.ts:83]: thanks for catching that

File: src/app/signers/data/UseSignerAddresses.ts (hunk 7 above)
Comment (BLuEScioN) [reply src/app/signers/data/UseSignerAddresses.ts:82]: done

File: src/app/signers/data/UseSignerAddresses.ts (hunk 8 above)
Comment (BLuEScioN) [reply src/app/signers/data/UseSignerAddresses.ts:86]: Pretty sure react query handles caching the paginated data internally through its pages and pageParams structure. Looking at the infinite query example from the docs:
The pagination parameters (pageParam, offset, limit) don't need to be in the queryKey because:
The pageParam is managed internally by the infinite query
Each "page" of data is stored in data.pages
The pagination state is tracked in data.pageParams
I really dont think it would be likely where we would have this exact same query key but with just a different limit

File: src/app/signers/data/useSigners.ts (hunk 10 above)
Comment (BLuEScioN) [reply src/app/signers/data/useSigners.ts:46]: See previous comment

File: src/app/signers/AddressesStackingCard.tsx [hunk 18]
```diff
@@ -3,32 +3,39 @@ import { UseQueryResult, useQueries, useQueryClient } from '@tanstack/react-quer
...
 import { Text } from '../../ui/Text';
 import { ExplorerErrorBoundary } from '../_components/ErrorBoundary';
 import { useSuspenseCurrentStackingCycle } from '../_components/Stats/CurrentStackingCycle/useCurrentStackingCycle';
 import { StatCardBase } from './StatsCardBase';
 import { SignersStackersData, useGetStackersBySignerQuery } from './data/UseSignerAddresses';
-import { useSuspensePoxSigners } from './data/useSigners';
+import { PoxSigner, useSuspensePoxSigners } from './data/useSigners';
 import { SignersStatsSectionSkeleton } from './skeleton';
 
 // TODO: This is currently not being used, but it may be in the future
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signers/AddressesStackingCard.tsx:17]: sure

File: src/app/signers/SignersTable.tsx (hunk 1 above)
Comment (BLuEScioN) [reply src/app/signers/SignersTable.tsx:307]: this is handled here 
export function formatSignerProposalMetric(metric: number): string {
  if (isNaN(metric)) return '-';
  if (metric === 0) return '0%';
  if (metric === 1) return '100%';
  return `${(metric * 100).toFixed(1)}%`;
}

File: src/app/signer/[signerKey]/AssociatedAddressesTable.tsx [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+}: {
+  children: ReactNode;
+  isLast: boolean;
+}) => {
+  return (
+    <Flex
+      alignItems={'center'}
+      gap={1}
+      py={6}
+      borderBottom={isLast ? 'none' : '1px solid var(--stacks-colors-borderSecondary)'}
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/AssociatedAddressesTable.tsx:36]: But now that I think about it, perhaps we can make that change in a separate PR to make the change cleaner

**Comment by BLuEScioN** [comment]
> image we might wanna make these cards same width on smaller screen (unless this is intentional)

Can you make a separate issue for this? This PR is for the signer id page

**Comment by BLuEScioN** [comment]
> image maybe we truncate the signer key even more on mobile to allow more width to the table?

Can you make a separate issue for this? This PR is for signer id page

File: src/app/signer/[signerKey]/AssociatedAddressesTable.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (He1DAr) [reply src/app/signer/[signerKey]/AssociatedAddressesTable.tsx:34]: for this one you can use:
borderBottom="1px solid var(--stacks-colors-borderSecondary)"
_last={{
  borderBottom: 'none',
}}
and it should work without the need for other changes

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (He1DAr) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx]: mind clarifying this? from what I understood it looks like missing will default to 0.05 if proposals_missed_count doesn't exist.

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+    px={6}
+    border="none"
+    sx={isFirst ? mobileBorderCss : {}}
+    width="fit-content"
+    position={isFirst ? 'sticky' : 'unset'}
+    left={0}
+    bg="surface"
+  >
+    <Flex
+      bg="hoverBackground"
```
[the remark this answers is not in the document]
Comment (He1DAr) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:48]: thoughts on this one

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+  }
+`;
+
+const Header = ({ headerTitle, isFirst }: { headerTitle: string; isFirst: boolean }) => (
+  <Th
+    py={3}
+    px={6}
+    border="none"
+    sx={isFirst ? mobileBorderCss : {}}
+    width="fit-content"
```
[the remark this answers is not in the document]
Comment (He1DAr) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:42]: thoughts on this one

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+import { Tr } from '../../../ui/Tr';
+import { ScrollableBox } from '../../_components/BlockList/ScrollableDiv';
+import { ExplorerErrorBoundary } from '../../_components/ErrorBoundary';
+import { useSuspenseCurrentStackingCycle } from '../../_components/Stats/CurrentStackingCycle/useCurrentStackingCycle';
+import { CycleFilter } from '../../signers/CycleFilter';
+import { mobileBorderCss } from '../../signers/consts';
+import { CycleSortFilter, CycleSortOrder } from './CycleSortFilter';
+import { StackingHistoryTableSkeleton } from './skeleton';
+
+const StyledTable = styled(Table)`
```
[the remark this answers is not in the document]
Comment (He1DAr) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:26]: thoughts on this one

[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:26]: Considering all these tables are going to get replaced by the end of the year, after I finished upgrading to Chakra v3 and merge in my table component, which will make the table styling and code consistent, I don't think it's a good use of time polishing temporary code

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:42]: see previous comment

File: src/app/signer/[signerKey]/StackingHistoryTable.tsx [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+  cycleid,
+  votingPower,
+  stxStacked,
+  latency,
+  approved,
+  rejected,
+  missing,
+}: {
+  index: number;
+  isFirst: boolean;
```
[the remark this answers is not in the document]
Comment (BLuEScioN) [reply src/app/signer/[signerKey]/StackingHistoryTable.tsx:104]: see previous comment
